### PR TITLE
Bug 1141093 - rhc snapshot save same filename didn't prompt conflict warning

### PR DESCRIPTION
**Pull request by dinhxuanvu**
The 'rhc snapshot save -f' command overwrites existing file without warning
if user inputs same filename for two different apps.

This commit modifies the rhc snapshow command to verify if the file already
exists and ask user for confirmation before proceeding.

Bug 1141093
Link <https://bugzilla.redhat.com/show_bug.cgi?id=1141093>

Signed-off-by: Vu Dinh <vdinh@redhat.com>

### CODE REVIEW

File: spec/rhc/commands/snapshot_spec.rb [hunk 1]
```diff
@@ -38,6 +38,30 @@
...
+        output.should match(/File #{@targz_filename} already exists. Do you want to overwrite this file (y\/n): /)
+        output.should match(/Pulling down a snapshot of application/)
+      end
+    end
+
+    context 'when failing to save a snapshot into existing file due to Interrupt' do
+      let(:arguments) {['snapshot', 'save', '--app', 'mockapp', '-f', 'mockapp.tar.gz', '-d']}
+      before do
+        File.stub(:exists?).and_return(true)
+        RHC::Commands::Snapshot.stub(STDIN.gets).and_return(Interrupt.new)
```
Comment (Miciah) [inline spec/rhc/commands/snapshot_spec.rb]: s/`return`/`raise`/

**Comment by dinhxuanvu** [comment]
@Miciah @tiwillia Finally! Ready for review...

File: lib/rhc/commands/snapshot.rb [hunk 2]
```diff
@@ -30,7 +30,12 @@ def save(app)
 
       raise RHC::DeploymentsNotSupportedException.new if options.deployment && !rest_app.supports?("DEPLOY")
 
-      filename = options.filepath ? options.filepath : "#{rest_app.name}.tar.gz"
+      if options.filepath and File.exists? options.filepath
+        return 0 unless RHC::Helpers.agree "File #{options.filepath} already exists. Do you want to overwrite this file? (yes|no): "
+        filename = options.filepath
+      else
+        filename = options.filepath ? options.filepath : "#{rest_app.name}.tar.gz"
+      end
```
Comment (Miciah) [inline lib/rhc/commands/snapshot.rb]: If the user does not specify the file path, so the default of `#{rest_app.name}.tar.gz` is chosen, and a file by that name exists, should we not print the same warning? Perhaps the logic should be more like the following:

```ruby
      filename = options.filepath || "#{rest_app.name}.tar.gz"
      if File.exists? filename
        return 0 unless RHC::Helpers.agree "yada yada"
      end
```

Comment (dinhxuanvu) [inline lib/rhc/commands/snapshot.rb]: I'm aware of that but my reason that since they choose the default name, it's likely that they want to overwrite the old one. However, I think your suggestion cleans up some codes as well so I may as well do it.
Thanks :+1:

File: spec/rhc/commands/snapshot_spec.rb [hunk 3]
```diff
@@ -38,6 +38,19 @@
       it { run_output.should_not match 'some save output' }
     end
 
+    context 'when saving a snapshot into existing file' do
+      let(:arguments) {['snapshot', 'save', '--app', 'mockapp', '-f', 'mockapp.tar.gz']}
+      before do
+        File.stub(:exists?).and_return(true)
+        RHC::Helpers.should receive(:agree).with(/Do you want to overwrite this file?/).and_return(true)
+        subject.class.any_instance.should_receive(:exec).with(/ssh #{@ssh_uri.user}@#{@ssh_uri.host} 'snapshot' > #{@app.name}.tar.gz/).and_return([0, 'some save output'])
```
Comment (Miciah) [inline spec/rhc/commands/snapshot_spec.rb]: Why is it `#{@app.name}.tar.gz` here when you use `mockapp.tar.gz` in the arguments array?

File: lib/rhc/commands/snapshot.rb [hunk 4]
```diff
@@ -30,7 +30,10 @@ def save(app)
 
       raise RHC::DeploymentsNotSupportedException.new if options.deployment && !rest_app.supports?("DEPLOY")
 
-      filename = options.filepath ? options.filepath : "#{rest_app.name}.tar.gz"
+      filename = options.filepath || "#{rest_app.name}.tar.gz"
+      if File.exists? filename
+        return 0 unless RHC::Helpers.agree "File #{options.filepath} already exists. Do you want to overwrite this file? (yes|no): "
```
Comment (Miciah) [inline lib/rhc/commands/snapshot.rb]: I think `options.filepath` should be changed to `filename` in the prompt.

**Comment by dinhxuanvu** [comment]
@tiwillia Ready for review after changing couple things to compliant @Miciah suggestion!

**Comment by tiwillia** [comment]
LGTM, lets [test] it!

**Comment by dinhxuanvu** [comment]
re-[test]

File: spec/rhc/commands/snapshot_spec.rb [hunk 5]
```diff
@@ -32,14 +32,26 @@
 
     context 'when saving a snapshot' do
       before do
+        RHC::Helpers.should_receive(:agree).with(/Do you want to overwrite this file?/).and_return(true)
```
Comment (Miciah) [inline spec/rhc/commands/snapshot_spec.rb]: Why is this needed? Are we assuming that `#{@app.name}.tar.gz` exists from a previous test?

Comment (dinhxuanvu) [inline spec/rhc/commands/snapshot_spec.rb]: The test apparently creates the file in advance so the file already exists:
`FileUtils.cp(File.expand_path('../../assets/targz_sample.tar.gz', __FILE__), @targz_filename)`

**Comment by Miciah** [comment]
Please [test] again now that #724 is merged!

**Comment by dinhxuanvu** [comment]
Looking good. Please merge @tiwillia  @Miciah

**Comment by Miciah** [comment]
openshift-bot, please [merge]!

**Comment by Miciah** [comment]
Thanks, openshift-bot!
